Update response and error API for resolve_customer

schaefi opened the pull request:
Align the implementation with the API documentation as provided by the SCC team

schaefi committed:
Update response and error API for resolve_customer

Align the implementation with the API documentation as
provided by the SCC team

schaefi commented:
@arharovets Hi Alex, thanks much for pointing me to your swagger documentation. With this PR I have updated the code to match with what you wrote there. There are a few things  that came to my mind though:

1. in AWS a service can be connected through other services. e.g lambda triggered by a gateway API. In this context it can happen that request or response data is provided or needed as base64 encoded. Because of that the API supports the `isBase64Encoded` attribute. This should not conflict with what you do because right now all responses will be plain json not encoded.
2. Due to 1. and also to stay compatible with response structures I have seen from cloud services so far, my response structure follows this pattern
```
{
    "statusCode": 200,
    "isBase64Encoded": false,
    "body": {
        ---follows-your-swagger-documented-response---
    }
}
```
Please let me know if this is ok for you which I hope. Thanks

3. Regarding HTTP status codes in case of errors. I adapted as much as I could with regards to your swagger docs. However, I like to prefer and handover the HTTP status code as we get it from the cloud service in case of an error according to the individual calls we make to the cloud. Please let me know if this is also ok for you. This might end up in a different status code for e.g the token expiration but will be compatible with the cloud and I believe this is important that we do not invent our own codes for this part.

For a summary of the request and response please also look at the README from here:

* https://github.com/SUSE-Enceladus/suse-saas-tools/blob/API_response_consolidation/aws/resolve_customer/README.rst

It has been updated in the context of this PR

Thanks much, we will get there :)

arharovets commented:
> Hi Alex, thanks much for pointing me to your swagger documentation. With this PR I have updated the code to match with what you wrote there.

Thank you a lot for including me here!

> There are a few things that came to my mind though:
> 
>     1. in AWS a service can be connected through other services. e.g lambda triggered by a gateway API. In this context it can happen that request or response data is provided or needed as base64 encoded. Because of that the API supports the `isBase64Encoded` attribute. This should not conflict with what you do because right now all responses will be plain json not encoded.
> 
>     2. Due to 1. and also to stay compatible with response structures I have seen from cloud services so far, my response structure follows this pattern
> 
> 
> ```
> {
>     "statusCode": 200,
>     "isBase64Encoded": false,
>     "body": {
>         ---follows-your-swagger-documented-response---
>     }
> }
> ```
> 
> Please let me know if this is ok for you which I hope. Thanks

Some feedback about this one:
- Does the `statusCode` always match the HTTP Status?
- "isBase64Encoded" added just for compatibility with requesters other than SCC?

>     3. Regarding HTTP status codes in case of errors. I adapted as much as I could with regards to your swagger docs. However, I like to prefer and handover the HTTP status code as we get it from the cloud service in case of an error according to the individual calls we make to the cloud. Please let me know if this is also ok for you. This might end up in a different status code for e.g the token expiration but will be compatible with the cloud and I believe this is important that we do not invent our own codes for this part.

I believe we should keep the status codes returned by the AWS but there's an overlap there:
`GetEntitlements` returns `InvalidParameterException` and `ThrottlingException`, both HTTP 400, while `ResolveCustomer` uses 400 for a wide range of token-related errors, like expired/missing token.

If I get what you're saying, you're suggesting we follow the HTTP codes and error messages returned by AWS. Is that correct? If yes, then we would need to expand the number of errors returned by the service to also follow Azure. Instead, we could do the same as we did with the API endpoints and consolidate errors. What do you think about this?

arharovets commented:
So, the response should look something like this:
```yaml
{
  "status": "200",
  "isBase64Encoded": false,
  "body": {
    "marketplaceIdentifier": "AWS | Azure",
    "marketplaceAccountId": "CustomerAWSAccountId",
    "customerIdentifier": "CustomerIdentifier",
    "entitlements": [
      {
        "customerIdentifier": "string",
        "productCode": "string",
        "expirationDate": 123123111231,
        "dimension": "string",
        "value": {
          "booleanValue": false,
          "doubleValue": 1,
          "integerValue": 2,
          "stringValue": "string"
        }
      }
    ]
  }
}
```

Please note these two attributes:
```yaml
"marketplaceIdentifier": "AWS | Azure",
"marketplaceAccountId": "CustomerAWSAccountId",
```

Ideally, responses should be marketplace-agnostic while identifying where the data came from, which is why attribute names are not AWS-specific. Do you think it is worth the brain cells to do this?

schaefi commented:
@arharovets

> Does the statusCode always match the HTTP Status?

To my knowledge yes. I take the status code from the AWS ClientError interface. This interface provides the status code as `ResponseMetadata -> HTTPStatusCode` It would be very weird if it's not an HTTP status

> "isBase64Encoded" added just for compatibility with requesters other than SCC

correct and if we maybe have to deal with data that needs a base64 encoding because of special characters, encoding issues, etc etc. I'd like the API to be prepared for such a situation without having the need to change the layout in an incompatible way

> If I get what you're saying, you're suggesting we follow the HTTP codes and error messages returned by AWS. Is that correct? If yes, then we would need to expand the number of errors returned by the service to also follow Azure. Instead, we could do the same as we did with the API endpoints and consolidate errors. What do you think about this?

I agree. If there is overlap in the error codes it would be in any case a good idea to categorize the error such that the codes becomes unique. For example one code that indicates a key expiration independent of the cloud service provider. I would not mess with the actual error message though and pass this just along as we get it. As of today I'm unclear about the different errors we need to classify individually. I will do some closer look and add at least one code point to handle the key expiration as it is listed in your swagger docs. If you agree that specific error classification can be added step by step on top of the default behavior which is passing along what we get from the CSP. Thoughts ?


> Ideally, responses should be marketplace-agnostic while identifying where the data came from, which is why attribute names are not AWS-specific. Do you think it is worth the brain cells to do this?

In the beginning I was wondering why we more or less replicate the attribute names but I think we will gain a good set of flexibility in doing that. So I like it, let's keep that

schaefi committed:
Allow custom error classification

Add classify_error method to allow individual HTTP status codes
when need be. Add CSP provided error code e.g AWS.ResolveCustomer.ExpiredTokenException
to show up in the error response too

schaefi commented:
@arharovets

> we could do the same as we did with the API endpoints and consolidate errors. 

I added a commit which now allows us to classify error codes individually and when needed. You can specify

```python
self.error = classify_error(
    self.error, 'AWS.ResolveCustomer.ExpiredTokenException', 400
)
```

Each AWS service maintains error codes as described in e.g 
https://docs.aws.amazon.com/marketplace/latest/APIReference/API_marketplace-metering_ResolveCustomer.html

So my code takes the default as it comes from the CSP but we can then individually assign unique status codes as we need them. As I said the actual error message stays untouched.

schaefi commented:
@arharovets alternatively we could also make the AWS error code a part of the error response, something like

```
{
    "isBase64Encoded": false,
    "statusCode": 400,
    "body": {
        "errors": {
            "Registration": "MESSAGE",
            "Exception": "AWS.ResolveCustomer.ExpiredTokenException"
        }
    }
}
```

Let me know what you prefer. If we go with status codes we will end up with non HTTP conform code numbers.

arharovets commented:
> > Does the statusCode always match the HTTP Status?
> 
> To my knowledge yes. I take the status code from the AWS ClientError interface. This interface provides the status code as `ResponseMetadata -> HTTPStatusCode` It would be very weird if it's not an HTTP status

If the values match, do we truly need the `statusCode` attribute in the response?

> > "isBase64Encoded" added just for compatibility with requesters other than SCC
> 
> correct and if we maybe have to deal with data that needs a base64 encoding because of special characters, encoding issues, etc etc. I'd like the API to be prepared for such a situation without having the need to change the layout in an incompatible way

I agree, thanks for clarifying!

> > If I get what you're saying, you're suggesting we follow the HTTP codes and error messages returned by AWS. Is that correct? If yes, then we would need to expand the number of errors returned by the service to also follow Azure. Instead, we could do the same as we did with the API endpoints and consolidate errors. What do you think about this?
> 
> I agree. If there is overlap in the error codes it would be in any case a good idea to categorize the error such that the codes becomes unique. For example one code that indicates a key expiration independent of the cloud service provider. I would not mess with the actual error message though and pass this just along as we get it. As of today I'm unclear about the different errors we need to classify individually. I will do some closer look and add at least one code point to handle the key expiration as it is listed in your swagger docs. If you agree that specific error classification can be added step by step on top of the default behavior which is passing along what we get from the CSP. Thoughts ?

Yes, let's keep it as is, with the HTTP codes and error messages coming from the marketplace, and then explore the categorization topic when it's time. Thanks!

> If we go with status codes we will end up with non HTTP conform code numbers.

Let's keep everything consistent (see above). SCC usually handles the errors based on their HTTP code, this way we know which error messages can be exposed to the end user, like validation errors for user input if the input is validated by the third party API and which cannot, like 500.

> > Ideally, responses should be marketplace-agnostic while identifying where the data came from, which is why attribute names are not AWS-specific. Do you think it is worth the brain cells to do this?
> 
> In the beginning I was wondering why we more or less replicate the attribute names but I think we will gain a good set of flexibility in doing that. So I like it, let's keep that

Great, thank you!

schaefi commented:
> Let's keep everything consistent (see above). 

ok done :)

> Yes, let's keep it as is, with the HTTP codes and error messages coming from the marketplace, and then explore the categorization topic when it's time. Thanks!

Great, with that agreement I will merge the PR and zip a new container for the cloud and update my test endpoint.

Thanks
